src/reentrancy.sw

```diff
@@ -17,24 +17,19 @@ pub fn is_reentrant() -> bool {
         fp: u64
     }
 
-    let mut caller_id = Option::None();
+    // Get our current contract ID
+    let current_id = contract_id();
 
     while internal {
         let saved_registers_pointer = get_saved_regs_pointer(call_frame_pointer);
-        let temp_caller_id = get_previous_caller_id(saved_registers_pointer)
-        // cleanup match syntax when match for enums lands
-        match caller_id {
-          Option::Some(_) => {
-              if Option::Some(temp_caller_id) == caller_id {
-                reentrancy = true;
-                internal = false;
-            } else {
-                internal = !caller_is_external();
-                call_frame_pointer = saved_registers_pointer + 48;
-            };
-          },
-          _ => {caller_id = Option::Some(temp_caller_id);}
-        }
+        let temp_id = get_previous_contract_id(saved_registers_pointer)
+        if temp_id == current_id {
+            reentrancy = true;
+            internal = false;
+        } else {
+            internal = !caller_is_external();
+            call_frame_pointer = saved_registers_pointer + 48;
+        };
     }
```

Comment:
Yep, I think this is better. :+1: 

The `call_frame_pointer` value might not be quite right on line 31, I think you need to dereference it.  i.e., at line 31 it's still a pointer to a pointer.

I think `get_current_frame_pointer()` and `get_previous_frame_pointer(current_fp)` functions would be good here, to keep it fairly readable.

Reply:
@otrho I have a few questions here.

1. RE: dereferencing. I'm sure it's simple (and obvious after the fact) but how exactly do I go about dereferencing the value on line 31? 
I understand the concept(eg. by using the `*` operator  in Rust ) but it's not clear to me how to do this in an `asm` block. On that note, as we have no docs on the usage of assemble, what would be some recommended reading on something similar? (ie: x86 assembly language? or something else...)

2. RE: a`get_current_frame_pointer()` function
I have this assignment already on line 16:
```
let mut call_frame_pointer = asm() {
        fp: u64
    };
```
My understanding is that `$fp` would always point to the call frame on top of the stack, so there should be no need to access this location more than once (ie: no need for a `get_current_frame_pointer()` function. I might have misunderstood though. (also, we do have a `contract_id()` function in the stdlib which returns `$fp` already...)

3. RE:  a `get_previous_frame_pointer(current_fp)` function
I do have 2 functions, which can be used together to basically achieve this.
`fn get_saved_regs_pointer(frame_ptr: u64)` & `fn get_previous_contract_id(saved_regs_ptr: u64)` 
should get us to the start of the previous call frame. I'm trying to use them to recover the `ContractId` stored at that location, so again, getting crystal clear on the differences between returning values and returning pointers(and when//how to do each!) is key for me to put this together correctly.

Reply:
1. Dereferencing would need to be done with an `lw`, which can now be done with ASM blocks... not the most elegant solution, but until we actually have reference types I think this is how we do it.
2. Yeah, you'd get the current `fp` once, and subsequent `fp`s are fetched from the saved registers array, until it's zero which indicates an external call and the bottom of the call stack.  I think that's right.  `contract_id()` actually reads the first field at `fp` out though, because the compiler adds that code -- I mentioned that recently somewhere on Slack.  Rather than returning `fp` itself.  But the code is deceptive... if the type of `contract_id()` was `u64` with the same ASM block, it _would_ read and return `fp`.
3. `get_previous_contract_id()` would be unhelpful again, as I assume it's similar to `contract_id()` and has a return type of `b256`.

Reply:
@otrho I've (hopefully) addressed your comments.
I opened a couple of new issues regarding the creation of some modules to help with accessing registers and opcodes via wrapper functions as well.
I'll work to add tests to this as well once the test harness is merged.